enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,354 @@
+
+#### Data model
+
+Two new top-level fields added to the record [data model][data_model]:
+
+* `structured` (type object): Original structured JSON log entry.
+* `schema` (type string): Schema name assigned by user.
+
+Relationship between `structured` and `message` fields:
+
+* EXACTLY ONE of `message` or `structured` MUST be present.
+* `schema` is OPTIONAL, defined and interpreted by the user.
+* The presence or absence of the `structured` and `message` fields can be used to detect structured/unstructured records.
+
+With an *Elasticsearch* store, `schema` is copied to the existing `viaq_index_name` field to set the index.
+`viaq_index_name` will continue to be used for compatibility and upgrade, but will eventually be removed.
+
+#### ClusterLogForwarder configuration
+
+The elements of the CLF API are:
+
+- *Input*: Selects logs, adds meta-data, creates _internal_ records.
+- *Pipeline*: Transforms, filters and routes internal records.
+- *Output*: Connects to consumers, encodes and sends _output_ records.
+
+New fields:
+
+`input.stuctured`: (string, default "JSON") Attempt to parse structured entries.
+The only legal value is "JSON", others may be added in future.
+
+An input record contains:
+
+* `structured` field if `structured=JSON` *and* JSON parse succeeds.
+* `message` field if `input.structured` is absent *or* JSON parse fails.
+
+`input.schema`: (string, optional) User-defined schema name.
+Adds a `schema` field to the internal record with a user-defined value.
+Not used by the forwarder, included in output record for the consumer.
+
+`pipeline.content`: (enum, default="Enhanced") Content type for messages
+
+* "Enhanced":
+  - if log entry parsed as JSON: meta-data + `strutured` object field.
+  - else: meta-data + `message` string field.
+* "Original": Original message string, no meta-data
+  - if log entry parsed as JSON: use original log entry as output record, no meta-data.
+  - else: `message` string field only, no meta-data
+* "Unstructured": meta-data + `message` string field, regardless of JSON parse.
+
+### User Stories
+
+#### I want a replacement for the defunct MERGE_JSON_LOG feature
+
+The "Structured" content format provides all the information formerly available via
+[MERGE_JSON_LOG][defunct_merge]. Log entry field `x` is available as `.structured.x`,
+instead of just `.x`
+
+**NOTE**: We could provide a "Merge" content format to exactly reproduce the [MERGE_JSON_LOG][defunct_merge] feature.
+We didn't do so because mixing [logging data-model][data-model] fields with user-defined fields can cause name clashes and invalid messages.
+We could revisit this if there is sufficient user demand, but it would be at the user's risk.
+
+```yaml
+inputs:
+- name: InputJSON
+  application: {}
+  structured: JSON
+outputs:
+- name: Output
+  type: fluentdForward
+pipelines:
+- inputRefs: [ InputJSON ]
+  outputRefs: [ Output ]
+```
+
+#### I want to index logs from applications with different schema in Elasticsearch
```

Comment:
Nit but do we need to explicitly call out "Elasticsearch" ?  Is it possible that any target could support a notion of "schema" (e.g. Cloudwatch).  Cloudwatch schema = logroup?

Reply:
Yes it is possible to use other notions of schema. Elasticsearch is a special case but important to solve given our past index explosion problems, so @sichvoge asked for this case to be explicitly spelled out.